### PR TITLE
[ClusterLegalizer] Cluster Legalizer API

### PR DESCRIPTION
**Pull request by AlexandreSinger**
The goal of this PR is to clean up the cluster legalization code in the packer (the code which creates the clusters and ensures that they are legal) into a clean API which can be called in an Analytical Placement flow which will be called externally to the Packer. In fact, the packer will never be called; so truly it would "replace" the packer.

Isolated the cluster legalization logic out of the packer's clustering algorithm in such a way that it can be used externally to the clusterer with no issue. Once the ClusterLegalizer API is constructed, it will allocate all of the memory that it needs for legalization; it can then be used to legalize clusters; and when it is destroyed it will remove all state created for clustering.

The cluster legalizer API maintains the legalized clusters internally as opposed to using the ClusterBlockId as an ID to the clusters. This allowed me to separate the ClusteredNetlist from the packing algorithm entirely. This importantly helped separate out all legalization logic from directly modifying the ClusteredNetlist generation code.

I did my best to remove as much global state as I could from the API to make it as self-contained as possible.

This change greatly cleaned up the packer code and removed many of its global variables (including the ClusterLegalizer itself, which is now a local variable!)

### CODE REVIEW

**Comment by AlexandreSinger** [comment]
This PR is still WIP since I have some small cleanups and documentation left; however, I wanted to get an idea on the test coverage. We are passing all of the regular tests and the NightlyTests are currently down, so I am just trying the VTR benchmarks for now. Below is the current results on the VTR benchmarks:

  | vtr_parse_results_prepack.txt | vtr_parse_results_legalizer.txt
-- | -- | --
vtr_flow_elapsed_time | 1 | 0.991681512
odin_synth_time |   |  
parmys_synth_time | 1 | 1.094871271
abc_depth | 1 | 1
abc_synth_time | 1 | 0.969154666
num_clb | 1 | 1
num_memories | 1 | 1
num_mult | 1 | 1
max_vpr_mem | 1 | 1.220798835
num_pre_packed_blocks | 1 | 1
num_post_packed_blocks | 1 | 1
device_grid_tiles | 1 | 1
pack_time | 1 | 1.00876255
placed_wirelength_est | 1 | 1
place_time | 1 | 0.976409566
placed_CPD_est | 1 | 1
min_chan_width | 1 | 1
routed_wirelength | 1 | 1
min_chan_width_route_time | 1 | 0.971295383
crit_path_routed_wirelength | 1 | 1
critical_path_delay | 1 | 1
geomean_nonvirtual_intradomain_critical_path_delay | 1 | 1
crit_path_route_time | 1 | 0.995582056

The routed wirelength, min channel width, and the CPD are all exactly the same (which means this API upgrade is functionally equivalent) which is very good news. The pack time did not change much, but the max memory appears to have increased by 22%.

@vaughnbetz This increase in max memory is what we discussed in our last meeting. This is likely caused by the old clusterer algorithm freeing all the intermediate clustering data after each cluster is created. The current design keeps all the intermediate clustering data (such as the intra-lb routing data) so that clusters can be modified later (even though it isnt). I think if we use your idea with a "clean cluster data" method, we may be able to bring down this max memory. Do you think that is necessary? It would be a "weird" interface since it implies that the cluster cannot be modified by the legalizer later if it was cleaned.

**Comment by vaughnbetz** [comment]
It would be good to get the memory increase on the largest Titan designs. A 20% or so memory increase isn't catastrophic, but it is noticable so it would be good to put the clustering data structures on a diet if this increase affects the biggest designs.

**Comment by AlexandreSinger** [comment]
@vaughnbetz I implemented what we discussed (a clean method). While looking into the data structures for the router data and (especially) the pb stats, I noticed that they take up a large amount of space. It makes sense to clean them when the user knows they will no longer be modifying a cluster. This problem is mainly caused by the clusterer storing data within the pb type itself, hopefully someday we can separate this out!

With this change, the max memory for the VTR benchmarks is now slightly better than it was before!

  | vtr_parse_results_prepack.txt | vtr_parse_results_legalizer.txt
-- | -- | --
vtr_flow_elapsed_time | 1 | 1.025408546
odin_synth_time |   |  
parmys_synth_time | 1 | 1.101305826
abc_depth | 1 | 1
abc_synth_time | 1 | 0.971543723
num_clb | 1 | 1
num_memories | 1 | 1
num_mult | 1 | 1
max_vpr_mem | 1 | 0.992619072
num_pre_packed_blocks | 1 | 1
num_post_packed_blocks | 1 | 1
device_grid_tiles | 1 | 1
pack_time | 1 | 1.008189595
placed_wirelength_est | 1 | 1
place_time | 1 | 1.046060717
placed_CPD_est | 1 | 1
min_chan_width | 1 | 1
routed_wirelength | 1 | 1
min_chan_width_route_time | 1 | 1.025420775
crit_path_routed_wirelength | 1 | 1
critical_path_delay | 1 | 1
geomean_nonvirtual_intradomain_critical_path_delay | 1 | 1
crit_path_route_time | 1 | 1.035469963

Here is the raw data in case you were curious.
[legalizer_comparison.xlsx](https://github.com/user-attachments/files/16841830/legalizer_comparison.xlsx)

For the VTR circuits, there appears to be no change in quality, runtime, or memory usage and now the API is significantly easier to work with.

I still have some small changes to make before I remove the WIP status, but I expect it to be done by tomorrow. We should wait for the NightlyTests though to merge this in. I am still on the ropes about what to do about the reclustering API; I will continue to think on it.

**Comment by vaughnbetz** [comment]
Great, thanks @AlexandreSinger

**Comment by AlexandreSinger** [comment]
@vaughnbetz I have finalized the PR and removed the WIP label. The API is now ready for review. I have decided to completely remove the re_clustering API since, while trying to upgrade it to use the new legalizer API, I found that most of the code would need to be either removed or completely re-thought. Since this API contains no CI tests, it would be challenging to upgrade it; especially since it would require me to test it in both a placement context (which the placer is currently not setup to handle performing cluster legalization) and a packing context (which is now trivial since the re_clustering API would literally just be a wrapper to this new legalizer). The code will still exist in the Git history if someone wants to bring the API back; but I do believe that a complete rewrite is in order to bring this feature fully in (but it would be much easier to rewrite than prior). Please let me know what you think about this.

I will run the Titan circuits this afternoon and see how this API performs. I expect the runtime, memory usage, and quality to be very similar.

**Comment by AlexandreSinger** [comment]
Here are the results on the Titan circuits:

  | titan_parse_results_master.txt | tan_parse_results_legalizer.txt
-- | -- | --
vtr_flow_elapsed_time | 1 | 0.981573162
num_LAB | 1 | 1
num_DSP | 1 | 1
num_M9K | 1 | 1
num_M144K | 1 | 1
max_vpr_mem | 1 | 0.993315751
num_pre_packed_blocks | 1 | 1
num_post_packed_blocks | 1 | 1
device_grid_tiles | 1 | 1
pack_time | 1 | 1.001249895
placed_wirelength_est | 1 | 1
place_time | 1 | 0.963296599
placed_CPD_est | 1 | 1
routed_wirelength | 1 | 1
critical_path_delay | 1 | 1
geomean_nonvirtual_intradomain_critical_path_delay | 1 | 1
crit_path_route_time | 1 | 0.992672691

QoR is identical, pack time is practically the same, max_vpr_mem is down slightly.

I also pulled out the pack memory to directly see how the memory of the packer was affected by these code improvements:

| | Geomean pack_mem |
| -- | --|
Master | 3798.451049 MiB
New | 3402.47893 MiB
% decrease | 10.43% decrease

For some of the largest circuits, this amounts to around a gigabyte of memory savings. I think even more can be saved if the data structures were organized a bit better.

Here is the raw data if you are curious:
[titan_legalizer_comparison.xlsx](https://github.com/user-attachments/files/16860517/titan_legalizer_comparison.xlsx)

**Review by soheilshahrouz — COMMENTED** [review]
I added a few comments to suggest minor improvements.

@AlexandreSinger, `cluster_util.cpp` was too large and with no clean structure. Your changes has significantly improved the readablity.

File: vpr/src/pack/pack.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/pack.cpp:73]: The original idea of this was to allow the prepacker object to be passed around in the global scope. The constructor would then do nothing and the init method would actually perform the prepacking. Now that the prepacker became a local variable this now looks strange; however, I do foresee this being useful in the future.

File: vpr/src/pack/pack.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/pack.cpp:204]: Yes. I noticed this too. But this PR is meant to be functionally equivalent to whatever was in the clusterer prior. I do not want to make any changes that may create a different output. I will put a todo above this line so it is documented.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:179]: You have a good point. Ill see what I can do.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1511]: That is actually common practice for asserts (at least I always did this in LLVM). Strings always resolve to true, so doing an assert like this will actually print a message when the assert is hit; which may make it easier to debug what is going on.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1647]: YES. This was the most painful bug to debug while working on this PR. That line of code used to be in the middle of the constructor; however I ran into the following problem:
1) The ClusterLegalizer object was created on the stack
2) The constructor was called
3) The t_ext_pin_util_targets constructor was called
4) t_ext_pin_util_targets crashed with VPR_FATAL_ERROR in the middle of it (which is expected for some testcases)
5) The ClusterLegalizer object is IMMEDIATELY destroyed (calling the destructor); since the stack is now being destroyed.
6) The destructor is called even though the object is not fully initialized causing a free of nullptr and crashing.

It was an absolute nightmare to debug and was challenging to fix in a clean way. I opted to just create the t_ext_pin_util_targets at the end of the constructor instead.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1547]: I have thought about that. I opted to just match whatever the Netlist class was doing.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:179]: I opted to use the same name, and use the "this->" syntax when it was confusing. I personally do not like the "_<member_variable>" syntax since I think "_<variable_name>" may be reserved by some compilers.

Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h]: Nevermind, the compiler was giving warnings. I will move to the `<member_variable>_` syntax.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (vaughnbetz) [reply vpr/src/pack/cluster_legalizer.cpp:1659]: I would mention the destructor is called if this fails (and VPR_FATAL_ERROR raises an exception) so we need to have the object in a state where the destructor will work. Hence do this last.

File: vpr/src/pack/cluster_util.cpp
Comment (vaughnbetz) [inline vpr/src/pack/cluster_util.cpp:1853]: It would be good to move this comment to the header file and make it Doxygen format.

File: vpr/src/pack/cluster_util.cpp
Comment (vaughnbetz) [inline vpr/src/pack/cluster_util.cpp:1966]: Move comment to header file.

Comment (vaughnbetz) [reply vpr/src/pack/cluster_util.cpp:1966]: update -> updates

**Review by vaughnbetz — APPROVED** [review]
Looks great, thanks!  I have a bunch of suggestions, but they're all around commenting etc. not the actual code implementation.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:304]: Yes. This call is ALWAYS needed if you want to generate a clustered netlist from the cluster legalizer. (technically you do not have to if you called "clean" on all clusters; however, this just becomes a NOOP in that case. It would be too complicated to mention that and not worth it at all).

I added more to the description to make it clear that this is required to generate a clustered netlist.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:464]: When I scanned the code, this is also what I observed. Its the max size of some queue which is used to store the number of "feasible blocks" in a given pb. Its a very strange structure and I plan to make the array dynamic in another PR later (or raise an issue to do so), since it really does not make sense.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1138]: This comment is referring to the context accesses. I want to avoid these "shadow" accesses to the global context because the I want to use this class outside of the packer. The hope is that the user will know what needs to be initialized before using this class.

File: vpr/src/pack/pack.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/pack.cpp:207]: I will raise an issue on this after the PR is merged.

File: vpr/src/pack/cluster.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster.cpp:387]: I will raise an issue after this PR is merged.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h]: Originally I wanted to allow clusters to be modified (removing a molecule from a cluster); however, I found that the feature was never used (and would never be tested). This should have been an unordered_set anyways. I turned it into a vector, since it is only used in the cluster_legalizer to destroy the cluster, which requires that we just iterate over the container.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:154]: That is correct. I have updated the example.

File: vpr/src/pack/cluster_legalizer.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.h:247]: I originally avoided this comment since a lot of the arguments I feel are unnecessary and should be cleaned up later (most of them are just to pre-allocate buffers to max values instead of just using dynamic arrays). I added comments for each param anyways and added a TODO to replace the static arrays for dynamic ones. Ill raise an issue after this PR is merged.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1371]: I will raise an issue after this PR is merged.

File: vpr/src/pack/cluster_legalizer.cpp
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_legalizer.cpp:1410]: I will raise an issue on this as well.

File: vpr/src/pack/cluster_util.h
Comment (AlexandreSinger) [reply vpr/src/pack/cluster_util.h:5]: I did the best I could. A lot of the methods actually did have documentation, but they were hidden within the implementation function itself. I have moved these to the header file and documented any of the methods I was confident I knew.

**Comment by AlexandreSinger** [comment]
@vaughnbetz Thank you so much for all of your comments. I really appreciate them. I have resolved most of them. I want to discuss with you forward-declaring the static methods; I personally do not like doing that, but its not a hill I will die on.

I am running nightly_test 1, 2, 3, and 4 tonight and will run 5, 6, and 7 tomorrow (they should be a bit quicker). I will let you know the results of these runs. Since this should be functionally equivalent, I do not expect golden results to change.

After this PR is merged, I plan to raise another PR fixing a limitation of this API. I will also raise a few issues based on your comments above which I did not resolve in this PR.

**Comment by vaughnbetz** [comment]
Thanks Alex. I'm not fanatical about declaring static routines at the top of the file (although I do prefer it) so happy to discuss.

**Pull request merged**
